### PR TITLE
Framework sketch

### PR DESCRIPTION
**Pull request by lwoodson**
A basic end-to-end sketch of the framework with interfaces defined but no hard implementations for extension points.   Grokking what is here should start with looking at the [project](https://github.com/devquixote/gumshoe/blob/framework_sketch/README.md) and [java readme](https://github.com/devquixote/gumshoe/blob/framework_sketch/java/README.md).  You can then probably make sense of everything else following the commits in order.  Further examples of use can be seen in the [integration tests](https://github.com/devquixote/gumshoe/tree/framework_sketch/java/src/test/java/com/bazaarvoice/gumshoe/integration).

Follow up work would be to add some missing tests and to write some implementations of `Publisher`, `Filter` and `Decorator` as described in the readme.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,7 +54,7 @@ for items in items_of(client):
   emit("did something")
   # do something else
   emit("did something else")
-  context().stop()
+  context().stop() # or could be context().fail()
```
Comment (aoudenne) [inline README.md:57]: At the top of this file, could we standardize "GumShoe" to "Gumshoe"?  I'm pretty sure it's all one word with no capital S).

Comment (aoudenne) [inline README.md:57]: Unless you really want it to be GumShoe.  I don't want to stifle your self-expression :)

Comment (lwoodson) [inline README.md:57]: Hahaha, no, I'll make the change.

**Comment by aoudenne** [comment]
This is the nicest code I've seen in forever.  Ship it!

**Comment by lwoodson** [comment]
Thanks for the review, @aoudenne

File: java/src/test/java/com/bazaarvoice/gumshoe/integration/InMemoryPublisher.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.bazaarvoice.gumshoe.integration;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import com.bazaarvoice.gumshoe.Publisher;
+
+public class InMemoryPublisher implements Publisher {
```
Comment (lewisren) [inline java/src/test/java/com/bazaarvoice/gumshoe/integration/InMemoryPublisher.java:9]: This doesn't seem very useful. Is this just for debugging?

**Comment by lewisren** [comment]
Rest LGTM

File: java/src/test/java/com/bazaarvoice/gumshoe/integration/InMemoryPublisher.java (hunk 2 above)
Comment (lwoodson) [inline java/src/test/java/com/bazaarvoice/gumshoe/integration/InMemoryPublisher.java:9]: Integration tests.  I can validate that the events I expect were published.  EX:  https://github.com/devquixote/gumshoe/blob/3d6a2007b2a5d5c732585f799bad165bda6435d5/java/src/test/java/com/bazaarvoice/gumshoe/integration/SingleThreadedIntegrationTest.java#L100-L116

**Comment by lwoodson** [comment]
Thanks @lewisren

**Pull request merged**